src/expertsystem/reaction/topology.py

```diff
@@ -336,42 +339,139 @@ def __eq__(self, other: object) -> bool:
                 return False
             if self.edges != other.edges:
                 return False
-            if self.edge_props != other.edge_props:
+            if any(
+                self.get_edge_props(i) != other.get_edge_props(i)
+                for i in self.edges
+            ):
                 return False
             if self.graph_node_properties_comparator is not None:
-                return self.graph_node_properties_comparator(
-                    self.node_props, other.node_props
+                return all(
+                    self.graph_node_properties_comparator(
+                        self.get_node_props(i), other.get_node_props(i)
+                    )
+                    for i in self.nodes
                 )
-            return self.node_props == other.node_props
+            return all(
+                self.get_node_props(i) == other.get_node_props(i)
+                for i in self.nodes
+            )
 
         raise NotImplementedError
 
-    def __copy__(self) -> "StateTransitionGraph[_EdgeType]":
-        """Makes a *shallow* copy.
+    @property
```

Comment:
Actually, why use composition?

Reply:
You mean why not inherit from topology? If thats the question, the idea was to make the Graph "frozen" from the topology point of view. So the add_node and edge functionality is not there anymore.

Reply:
Yes, previously `StateTransitionGraph` inherited from `Topology`. But sounds good 👍